src/execution/operator/csv_scanner/table_function/csv_multi_file_info.cpp

```diff
@@ -0,0 +1,407 @@
+	                                    csv_names, csv_types, csv_data.csv_schema, gstate.SingleThreadedRead(), nullptr,
+	                                    false);
+}
+
+shared_ptr<BaseFileReader> CSVMultiFileInfo::CreateReader(ClientContext &context, GlobalTableFunctionState &gstate_p,
+                                                          const string &filename, idx_t file_idx,
+                                                          const MultiFileBindData &bind_data) {
+	auto &gstate = gstate_p.Cast<CSVGlobalState>();
+	auto &csv_data = bind_data.bind_data->Cast<ReadCSVData>();
+
+	auto options = csv_data.options;
+	if (bind_data.file_list->GetExpandResult() == FileExpandResult::SINGLE_FILE) {
+		options.auto_detect = false;
+	}
+	shared_ptr<CSVBufferManager> buffer_manager;
+	if (file_idx == 0) {
+		buffer_manager = csv_data.buffer_manager;
+		if (buffer_manager && buffer_manager->GetFilePath() != filename) {
+			buffer_manager.reset();
+		}
+	}
+	return make_shared_ptr<CSVFileScan>(context, filename, std::move(options), bind_data.file_options, bind_data.names,
+	                                    bind_data.types, csv_data.csv_schema, gstate.SingleThreadedRead(),
+	                                    std::move(buffer_manager), false);
+}
+
+shared_ptr<BaseFileReader> CSVMultiFileInfo::CreateReader(ClientContext &context, const string &filename,
+                                                          CSVReaderOptions &options,
+                                                          const MultiFileReaderOptions &file_options) {
+	return make_shared_ptr<CSVFileScan>(context, filename, options, file_options);
+}
+
+shared_ptr<BaseUnionData> CSVMultiFileInfo::GetUnionData(shared_ptr<BaseFileReader> scan_p, idx_t file_idx) {
+	auto &scan = scan_p->Cast<CSVFileScan>();
+	auto data = make_shared_ptr<CSVUnionData>(scan_p->GetFileName());
+	if (file_idx == 0) {
+		data->options = scan.options;
+		data->names = scan.GetNames();
+		data->types = scan.GetTypes();
+		data->reader = std::move(scan_p);
+	} else {
+		data->options = std::move(scan.options);
+		data->names = scan.GetNames();
+		data->types = scan.GetTypes();
+	}
+	data->options.auto_detect = false;
+	return data;
+}
+
+void CSVMultiFileInfo::FinalizeReader(ClientContext &context, BaseFileReader &reader) {
+	auto &csv_file_scan = reader.Cast<CSVFileScan>();
+	csv_file_scan.InitializeFileNamesTypes();
+	csv_file_scan.SetStart();
+}
+
+bool CSVMultiFileInfo::TryInitializeScan(ClientContext &context, shared_ptr<BaseFileReader> &reader,
+                                         GlobalTableFunctionState &gstate_p, LocalTableFunctionState &lstate_p) {
+	auto &gstate = gstate_p.Cast<CSVGlobalState>();
+	auto &lstate = lstate_p.Cast<CSVLocalState>();
+	auto csv_reader_ptr = shared_ptr_cast<BaseFileReader, CSVFileScan>(reader);
+	gstate.FinishScan(std::move(lstate.csv_reader));
+	lstate.csv_reader = gstate.Next(csv_reader_ptr);
```

Comment:
Is this thread safe? Next defines the boundaries of a scan, but I haven't found its lock.

Reply:
`CSVMultiFileInfo::TryInitializeScan` is called under the multi file reader lock (see `TryInitializeNextBatch`)